### ISSUE TITLE
one-var and no-mixed-requires conflict

### ISSUE THREAD

**Issue by drewed** [eslint/eslint#3091]
Given the following rule configuration:
```json
{
    "rules": {
        "one-var": [2, "always"],
        "no-mixed-requires": [2, false]
    }
}
```

I've got a block of code that can't satisfy eslint (v.0.24.1) as follows:

```javascript
const locale = this.cookies.get("_lang"),
    {messages} = require(`data/${locale}`); // <-- triggers no-mixed-requires
```

Separating the require call from the locale definition also errors:

```javascript
const locale = this.cookies.get("_lang");

const {messages} = require(`data/${locale}`); // <-- triggers one-var
```

It seems like there should be a way to get the two rules to play nicely with each other.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by xjamundx** [comment]
I think this is an interesting problem. What would you like to see? (for me I just use es6 `import`, which you **can't** mix wiht a `var` block), but I'm not sure what other folks would like to see

**Comment by lo1tuma** [comment]
Looks like a duplicate of #1232 and #2009.

**Comment by drewed** [comment]
@xjamundx I'd like to see the no-mixed-requires take precedence over the one-var rule such that one-var would request duplicate rules be lumped into the appropriate declaration block while still allowing the two groups to be separated which would be different from simply disabling one or the other rule as indicated in #1232.  In other words, one-var should be aware of the state of no-mixed-requires.  That or the documentation should indicate that the two rules are incompatible.

**Comment by ilyavolodin** [comment]
@drewed Every ESLint rule is a plugin, so none of the rules are aware of the state of other rules. We could possibly add an option to `one-var` that would ignore variables that have a value of require function, but we can't make `one-var` behave differently depending on the state of `no-mixed-require`

**Comment by drewed** [comment]
@ilyavolodin Makes sense.  One wouldn't want to over-complicate the system with cross-dependencies.

**Comment by nzakas** [comment]
I think this might be one of those situations where you need to manually disable one-var around areas, or possibly create a custom rule. I'm not sure special casing requires in one-var makes sense.

**Comment by xjamundx** [comment]
I'm pretty sure this issue is about to get closed, but one idea I had would be a `functions-only` version of `one-var`. Where it only applies inside of functions and never at the root level (including globals, es6 modules or node modules). But I personally don't really worry that much about either of these rules, because I use `let` and `import`, so it doesn't really apply and I don't want to champion any big change.

**Comment by drewed** [comment]
@nzakas That would be a work around but shouldn't be a long term solution as it creates a poor user experience to have to disable eslint around "proper syntax" (in terms of no-mixed-require).

I will understand if the decision is made to not support the two rules used in tandem in which case I urge for the documentation to reflect that the two rules should not be used simultaneously.

**Comment by nzakas** [comment]
@drewed Unfortunately, we can't support every variation of code that people come to us with - it's just not scalable. That's why we have the inline comments to disable around particular patterns as a way of working around that fact.

In this case, though, I'm not sure I understand what you're trying to accomplish. It seems like you definitely want one `var` statement only. Given that, why do you also have `no-mixed-requires` enabled? `no-mixed-requires` specifically forces you to separate out into multiple declarations when you have both `require` and non-`require` declarations.

Can you explain more about what your expectations are?

**Comment by drewed** [comment]
@nzarkas The intent of the `one-var` rule is to consolidate variable declarations scattered throughout a single scope to one place which generally makes for cleaner code (in my opinion).  The `no-mixed-requires` directs that require calls assigned to values should not be mixed with other variable declarations which can help set apart external dependencies from internal data.  What I would like to be able to do would be something one could potentially describe as `two-var`.  I'd like the declarations consolidated and organized at the top of the scope in such a way that require calls are separate from local variables but neither are scattered throughout the scope of any given block.

**Comment by xjamundx** [comment]
@drewed I think you propose an interesting solution and I'd be happy to create a `two-var` plugin if you want to work on that. It would basically combine both rules into a single one.

**Comment by drewed** [comment]
@xjamundx What help would you like me to provide?

**Comment by xjamundx** [comment]
Well basically I'm saying. I don't think this is going to go into eslint at this time, but if you're interested in creating a plugin to provide the functionality I'd be happy to offer support.

**Comment by drewed** [comment]
If you're willing to point me in the right direction, I'm happy to dive in.  I've forked the repo and setup the dev environment.

**Comment by xjamundx** [comment]
Okay @drewed any chance you can join https://gitter.im/eslint/eslint and I'll work with you directly (and hopefully we can close this for now)

**Comment by xjamundx** [comment]
We're going to work on a plugin https://github.com/drewed/eslint-plugin-twovar will post an update here once it's done

**Comment by nzakas** [comment]
I think @xjamundx's assessment is correct here - you're asking for a setting that can't rationally exist in ESLint between these two rules, so a plugin is definitely the better way to go.

Closing this issue so we don't track it, but feel free to post updates.
